blockstore.go

```diff
@@ -143,24 +336,47 @@ func (b *Blockstore) GetSize(cid cid.Cid) (int, error) {
 		}
 		return err
 	})
-	if lmdb.IsNotFound(err) {
+	switch {
+	case err == nil:
+		return size, nil // shortcircuit the happy path with no comparisons.
+	case lmdb.IsNotFound(err) || lmdb.IsErrno(err, lmdb.BadValSize):
 		err = blockstore.ErrNotFound
 	}
 	return size, err
 }
 
 func (b *Blockstore) Put(block blocks.Block) error {
-	return b.env.Update(func(txn *lmdb.Txn) error {
+	b.oplock.RLock()
+	defer b.oplock.RUnlock()
+
+Retry:
+	err := b.env.Update(func(txn *lmdb.Txn) error {
 		err := txn.Put(b.db, block.Cid().Hash(), block.RawData(), lmdb.NoOverwrite)
 		if err == nil || lmdb.IsErrno(err, lmdb.KeyExist) {
 			return nil
 		}
 		return err
 	})
+	if lmdb.IsMapFull(err) {
+		o := b.dedupGrow   // take the deduplicator under the lock.
+		b.oplock.RUnlock() // drop the concurrent lock.
+		var err error
+		o.Do(func() { err = b.grow() })
+		if err != nil {
+			return fmt.Errorf("lmbd put failed: %w", err)
+		}
+		b.oplock.RLock() // reclaim the concurrent lock.
+		goto Retry
```

Comment:
I applaud the use of goto! just don't let steb see it :)

Reply:
It does work very nicely here.